File: catalog_app/src/types/manifest.ts
export interface ManifestMetadata {
  project_name: string
  env: string
  schema: string
  connection_type: string
  generated_at: string
}

export interface ManifestNode {
  name: string
  path: string
  raw_sql: string
  compiled_sql: string
  ddl: string
  target: string
  materialization: string
  refs: string[]
  sources: [string, string][]
  description: string
  tags: string[]
  enabled: boolean
}

export interface ManifestSource {
  schema: string
  database: string | null
  tables: string[]
}

export interface Manifest {
  metadata: ManifestMetadata
  nodes: Record<string, ManifestNode>
  sources: Record<string, ManifestSource>
  parent_map: Record<string, string[]>
  child_map: Record<string, string[]>
  batches: string[][]
}

export interface TestResultEntry {
  model: string
  column: string
  test_type: string
  severity: string
  passed: boolean
  params?: Record<string, unknown>
  where?: string
  failures_count?: number
  failures_sample?: string[][]
  error?: string
}

export interface TestResultsSummary {
  total: number
  passed: number
  failed: number
  warned: number
  errored: number
}

export interface TestResults {
  metadata: {
    env: string
    schema: string
    generated_at: string
  }
  summary: TestResultsSummary
  results: TestResultEntry[]
}

File: python/qraft/manifest.py
import json
from collections import defaultdict
from datetime import datetime, timezone
from pathlib import Path

from qraft import _core
from qraft.config.models import EnvConfig, Model


def generate_manifest(
    compiled_batches: list[list[_core.CompiledModel]],
    model_map: dict[str, Model],
    env: EnvConfig,
    dag: _core.DagHandle,
    batches: list[list[str]],
    project_name: str,
    target_dir: Path,
) -> dict:
    """Build manifest dict and write to target/manifest.json."""

    # Build parent_map and child_map from DAG edges
    edges = _core.dag_edges(dag)
    parent_map: dict[str, list[str]] = defaultdict(list)
    child_map: dict[str, list[str]] = defaultdict(list)
    for parent, child in edges:
        parent_map[child].append(parent)
        child_map[parent].append(child)

    # Add source references to parent_map
    for batch in compiled_batches:
        for compiled in batch:
            for source_name, table_name in compiled.sources:
                source_key = f"source:{source_name}.{table_name}"
                if source_key not in parent_map[compiled.name]:
                    parent_map[compiled.name].append(source_key)

    # Build nodes dict from compiled models
    nodes = {}
    for batch in compiled_batches:
        for compiled in batch:
            model = model_map.get(compiled.name)
            nodes[compiled.name] = {
                "name": compiled.name,
                "path": str(model.path) if model else f"{compiled.name}.sql",
                "raw_sql": model.raw_sql if model else "",
                "compiled_sql": compiled.compiled_sql,
                "ddl": compiled.ddl,
                "target": compiled.target,
                "materialization": compiled.materialization,
                "refs": compiled.refs,
                "sources": compiled.sources,
                "description": compiled.description,
                "tags": compiled.tags,
                "enabled": compiled.enabled,
            }

    # Build sources dict from env config
    sources = {}
    for source_name, src_config in env.sources.items():
        sources[source_name] = {
            "schema": src_config.schema,
            "database": src_config.database,
            "tables": src_config.tables,
        }

    manifest = {
        "metadata": {
            "project_name": project_name,
            "env": env.name,
            "schema": env.schema,
            "connection_type": env.connection.type,
            "generated_at": datetime.now(timezone.utc).isoformat(),
        },
        "nodes": nodes,
        "sources": sources,
        "parent_map": dict(parent_map),
        "child_map": dict(child_map),
        "batches": batches,
    }

    out_path = target_dir / "manifest.json"
    out_path.parent.mkdir(parents=True, exist_ok=True)
    out_path.write_text(json.dumps(manifest, indent=2), encoding="utf-8")
    return manifest


def load_manifest(target_dir: Path) -> dict:
    """Read manifest.json from target/."""
    return json.loads(
        (target_dir / "manifest.json").read_text(encoding="utf-8")
    )
